vm/systemSmartContracts/staking_test.go

```diff
@@ -1 +1,448 @@
 package systemSmartContracts
+
+import (
+	"bytes"
+	"encoding/json"
+	"math/big"
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go/vm"
+	"github.com/ElrondNetwork/elrond-go/vm/mock"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+	"github.com/stretchr/testify/assert"
+)
+
+func CreateVmContractCallInput() *vmcommon.ContractCallInput {
+	return &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  []byte("tralala1"),
+			Arguments:   nil,
+			CallValue:   big.NewInt(0),
+			GasPrice:    big.NewInt(0),
+			GasProvided: big.NewInt(0),
+			Header:      &vmcommon.SCCallHeader{},
+		},
+		RecipientAddr: []byte("tralala2"),
+		Function:      "something",
+	}
+}
+
+func TestNewStakingSmartContract_NilStakeValueShouldErr(t *testing.T) {
+	t.Parallel()
+
+	eei := &mock.SystemEIStub{}
+	stakingSmartContract, err := NewStakingSmartContract(nil, eei)
+
+	assert.Nil(t, stakingSmartContract)
+	assert.Equal(t, vm.ErrNilInitialStakeValue, err)
+}
+
+func TestNewStakingSmartContract_NilSystemEIShouldErr(t *testing.T) {
+	t.Parallel()
+
+	stakeValue := big.NewInt(100)
+	stakingSmartContract, err := NewStakingSmartContract(stakeValue, nil)
+
+	assert.Nil(t, stakingSmartContract)
+	assert.Equal(t, vm.ErrNilSystemEnvironmentInterface, err)
+}
+
+func TestNewStakingSmartContract_NegativeStakeValueShouldErr(t *testing.T) {
+	t.Parallel()
+
+	stakeValue := big.NewInt(-100)
+	eei := &mock.SystemEIStub{}
+	stakingSmartContract, err := NewStakingSmartContract(stakeValue, eei)
+
+	assert.Nil(t, stakingSmartContract)
+	assert.Equal(t, vm.ErrNegativeInitialStakeValue, err)
+}
+
+func TestNewStakingSmartContract(t *testing.T) {
+	t.Parallel()
+
+	stakeValue := big.NewInt(100)
+	eei := &mock.SystemEIStub{}
+	stakingSmartContract, err := NewStakingSmartContract(stakeValue, eei)
+
+	assert.NotNil(t, stakingSmartContract)
+	assert.Nil(t, err)
+}
+
+func TestStakingSC_ExecuteInit(t *testing.T) {
+	t.Parallel()
+
+	stakeValue := big.NewInt(100)
+	eei, _ := NewVMContext(&mock.BlockChainHookStub{}, &mock.CryptoHookStub{})
+	eei.SetSCAddress([]byte("addr"))
```

Comment:
`eei.SetSCAddress()` is a step that needs to be performed on the EEI before calling any function of a SC. Because the EEI will only allow a SmartContract to access its own state exclusively, the EEI must be informed of the address of the SmartContract that is to be called. When the SmartContract later calls `GetStorage()`, the EEI will retrieve the storage under that specific SC address. This is why calling `eei.SetSCAddress()` happens *before* the SC is called. Normally, preparing the EEI is a step performed by the VM automatically, when a SC call is requested.

Reply:
ok